_posts/2022-08-17-using-the-kani-rust-verifier-on-tokio-bytes.md

```diff
@@ -0,0 +1,556 @@
+Let's begin with a property that should hold regardless of the representation: the instance's length must be within bounds of its capacity.
+We will also add calls to appropriate methods, to be defined, for each representation kind.
+In Kani, we can write a method `is_valid` to express when a type `T` is well-formed:
+
+```rust
+impl BytesMut {
+    fn is_valid(&self) -> bool {
+        self.len <= self.cap
+            && match self.kind() {
+                KIND_VEC => self.is_valid_kind_vec(),
+                KIND_ARC => self.is_valid_kind_arc(),
+                _ => false,
+            }
+    }
+}
+```
+
+For `KIND_VEC` we need to know a bit more about how a `BytesMut` is constructed.
+The implementation, which we've slightly simplified (just in this post), is as follows.
+Internally the constructor allocates the backing buffer inside a `Vec<u8>` with sufficient capacity.
+The `BytesMut` instance is then built out of the fields of the vector.
+
+```rust
+impl BytesMut {
+    pub fn with_capacity(capacity: usize) -> BytesMut {
+        BytesMut::from_vec(Vec::with_capacity(capacity)) //< (1)
+    }
+
+    pub(crate) fn from_vec(mut vec: Vec<u8>) -> BytesMut {
+        let ptr = vptr(vec.as_mut_ptr());                //< (2)
+        let len = vec.len();                             //<
+        let cap = vec.capacity();                        //<
+        mem::forget(vec);                                //< (3)
+
+        let data = KIND_VEC;                             //< (4)
+
+        BytesMut {
+            ptr,
+            len,
+            cap,
+            data: data as *mut _,
+        }
+    }
+}
+```
+
+In the above, we've annotated the steps:
+1. A `Vec<u8>` is used to allocate a buffer of the requested capacity
+2. Set `ptr` to point to the `vec`'s buffer ([`as_mut_ptr()`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.as_mut_ptr) returns the raw pointer (`*mut u8`) and `vptr()`, an internal function, turns this into a `NonNull<u8>`)
+3. Use [`mem::forget`](https://doc.rust-lang.org/std/mem/fn.forget.html) to ensure that `vec`'s destructor will *not* run when `from_vec` returns (notice that `vec` is moved into `from_vec`).
+This is essential to ensure that the buffer pointed to by `ptr` remains valid.
+4. Set `data` to `KIND_VEC` (`=0b1`), which is subsequently cast to a raw pointer `*mut Shared`
+
+This is interesting, but unfortunately there is not a great deal more we can say for `is_valid_kind_vec()`.
+If we have an instance using `KIND_VEC` then we know that the `ptr` and `cap` must be related to the "original vector".
+For example, immediately after construction, we know `ptr` must be exactly `vec`'s buffer and similarly for `cap`.
+However, we no longer have access to this `vec`, so for now we will make our well-formedness check a no-op.
+[Don't worry, we will return to this later!]
+
+```rust
+fn is_valid_kind_vec(&self) -> bool {
+    assert!(self.kind() == KIND_VEC);
```

Comment:
Should this be an `assert!` or `return false`

Reply:
There is `match` statement on the higher-level `is_valid` method, so this assert is redundant. It's kept later when `is_valid_kind_vec` is extended, and the assert helps to avoid usage from a generic `BytesMut` object, so I would keep it.